vmdb/app/models/ems_refresh/save_inventory_helper.rb

```diff
@@ -32,23 +34,19 @@ def save_inventory_prep(find_key, child_keys, extra_keys)
     return find_key, child_keys, extra_keys, remove_keys
   end
 
-  def save_inventory(type, klass, parent, hash, deletes, new_records, record_index, record_index_columns, find_key, child_keys, remove_keys)
-    # Backup keys that cannot be written directly to the database
-    key_backup = backup_keys(hash, remove_keys)
-
+  def save_inventory(type, parent, hash, deletes = nil, new_records = nil, record_index = nil,
+                     record_index_columns = nil, find_key = nil)
     # Find the record, and update if found, else create it
     found = find_key.blank? ? parent.send(type) : self.save_inventory_record_index_fetch(record_index, record_index_columns, hash, find_key)
     if found.nil?
-      found = klass.new(hash)
-      new_records.nil? ? parent.send("#{type}=", found) : new_records << found
+      found = find_key ? parent.send(type).build(hash) : parent.send("build_#{type}", hash)
+      # TODO: these should not be necessary. Probably missing some parent.save calls
```

Comment:
> Why are we doing children(true) everywhere?

The `children(true)` is done by the caller to get a set of records "to be deleted", which most of the time is all of them.  That array is passed into save_inventory along with an Array of Hashes of the records we expect to be in the database.  So, as we loop through the Hashes, we determine if they are new records or updates.  If it's an update we remove that existing record from the "to be deleted" Array, so in the end, everything left it that Array is what we have to actually delete from the database.

The `(true)` was to ensure that the in memory cache was ignored in case it was cached differently ahead of time (I can't remember specifics, but I think there were actual problems where that occurred, but perhaps that can be removed now).

@kbrock wishes to bury this into the save_inventory code in #1340 

> good news: save_inventory only has 2 callers: save_inventory_single and save_inventory_multi.

Oh, right... so, save_inventory is basically a DRY method for those two methods to deal with most of the commonality, particularly, `save_inventory` is for building a single record.  However, save_inventory_multi's purpose is to efficiently update a has_many.  As such, we don't want to save record by record in save_inventory.  Instead, the save_inventory just returns all of the new_records as an Array for batch insert.

Incidentally, to get even more out of batching, I really didn't want the `update_attributes!` in save_inventory when the record already exists.  I was hoping I could just update the attributes of the existing record in memory and then call `.save!` on the parent, but saving the parent doesn't save updated children.

> ```
>    new_records << p.children.build(name: "c1")
>    new_records << p.children.build(name: "c2")
>
>    # p.children.push(new_records) # works with or without this line
>    p.save
> ```

So the difference here is you are building and then pushing, so the build does the association linking for you and yes, the push is not needed.  That's not what we did in the original code.  In the original code code we did a .new and collected up all the new records into an array.  Then it deletes the records to be deleted in one shot and pushes the new records to be added in another shot.  Thus AR does a batch delete (one query) followed by a batch insert (one query per insert, but all in one transaction, so only 1 BEGIN/COMMIT pair).  The push is nice because it writes the batch to the database immediately if the parent is already saved.  Otherwise, if it's a new record it will delay writing until the parent is saved, hopefully creating a bigger transaction batch by including the parent and all other unsaved associations in the same BEGIN/COMMIT pair.

Example of batch push:
```ruby
** master - vmdb_development on localhost
Loading development environment (Rails 3.2.17)
[1] pry(main)> v = Vm.first
  Vm Load (5.2ms)  SELECT "vms".* FROM "vms" WHERE "vms"."template" = 'f' AND "vms"."type" IN ('Vm', 'VmInfra', 'VmKvm', 'VmMicrosoft', 'VmRedhat', 'VmVmware', 'VmXen', 'VmCloud', 'VmAmazon', 'VmOpenstack') LIMIT 1
  Vm Inst Including Associations (108.3ms - 1rows)
=> #<VmVmware id: 3, vendor: "vmware", format: nil, version: nil, name: "GM CentOS Trunk Appl - Retirement Testing", description: nil, location: "GM CentOS Trunk Appl - Retiremen/GM CentOS Trunk Ap...", config_xml: nil, autostart: nil, host_id: 2, last_sync_on: nil, created_on: "2014-11-24 03:08:25", updated_on: "2015-01-16 20:41:15", storage_id: 2, guid: "27f5118c-7387-11e4-b538-3c15c2c7fbb8", ems_id: 1, last_scan_on: nil, last_scan_attempt_on: nil, uid_ems: "422fb102-72d6-3604-48b5-9d014318dd45", retires_on: nil, retired: nil, boot_time: nil, tools_status: "toolsNotRunning", standby_action: "checkpoint", power_state: "off", state_changed_on: "2014-11-24 03:08:25", previous_state: nil, connection_state: "connected", last_perf_capture_on: nil, blackbox_exists: nil, blackbox_validated: nil, registered: nil, busy: nil, smart: nil, memory_reserve: 0, memory_reserve_expand: false, memory_limit: -1, memory_shares: 81920, memory_shares_level: "normal", cpu_reserve: 0, cpu_reserve_expand: false, cpu_limit: -1, cpu_shares: 4000, cpu_shares_level: "normal", cpu_affinity: nil, ems_created_on: nil, template: false, evm_owner_id: nil, ems_ref_obj: "--- !ruby/string:VimString\nstr: vm-101616\nxsiType: ...", miq_group_id: nil, linked_clone: false, fault_tolerance: false, type: "VmVmware", ems_ref: "vm-101616", ems_cluster_id: 3, retirement_warn: nil, retirement_last_warn: nil, vnc_port: nil, flavor_id: nil, availability_zone_id: nil, cloud: false, retirement_state: nil, cloud_network_id: nil, cloud_subnet_id: nil, cloud_tenant_id: nil, raw_power_state: "poweredOff", publicly_available: nil, orchestration_stack_id: nil>
[2] pry(main)> v.guest_applications(true)
  GuestApplication Load (4.4ms)  SELECT "guest_applications".* FROM "guest_applications" WHERE "guest_applications"."vm_or_template_id" = 3
  GuestApplication Inst Including Associations (0.0ms - 0rows)
=> []
[3] pry(main)> v.guest_applications.push([GuestApplication.new, GuestApplication.new])
   (0.2ms)  BEGIN
  SQL (2.3ms)  INSERT INTO "guest_applications" ("arch", "description", "host_id", "language", "name", "package_name", "path", "product_icon", "product_key", "release", "transform", "typename", "vendor", "version", "vm_or_template_id") VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15) RETURNING "id"  [["arch", nil], ["description", nil], ["host_id", nil], ["language", nil], ["name", nil], ["package_name", nil], ["path", nil], ["product_icon", nil], ["product_key", nil], ["release", nil], ["transform", nil], ["typename", nil], ["vendor", nil], ["version", nil], ["vm_or_template_id", 3]]
  SQL (0.3ms)  INSERT INTO "guest_applications" ("arch", "description", "host_id", "language", "name", "package_name", "path", "product_icon", "product_key", "release", "transform", "typename", "vendor", "version", "vm_or_template_id") VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15) RETURNING "id"  [["arch", nil], ["description", nil], ["host_id", nil], ["language", nil], ["name", nil], ["package_name", nil], ["path", nil], ["product_icon", nil], ["product_key", nil], ["release", nil], ["transform", nil], ["typename", nil], ["vendor", nil], ["version", nil], ["vm_or_template_id", 3]]
   (0.4ms)  COMMIT
=> [#<GuestApplication id: 1, name: nil, vendor: nil, version: nil, description: nil, package_name: nil, product_icon: nil, transform: nil, language: nil, typename: nil, vm_or_template_id: 3, product_key: nil, path: nil, arch: nil, host_id: nil, release: nil>,
 #<GuestApplication id: 2, name: nil, vendor: nil, version: nil, description: nil, package_name: nil, product_icon: nil, transform: nil, language: nil, typename: nil, vm_or_template_id: 3, product_key: nil, path: nil, arch: nil, host_id: nil, release: nil>]
```

Example of new record push and subsequent save:
```ruby
[4] pry(main)> v = VmVmware.new(:name => "x", :location => "x", :vendor => "vmware")
=> #<VmVmware id: nil, vendor: "vmware", format: nil, version: nil, name: "x", description: nil, location: "x", config_xml: nil, autostart: nil, host_id: nil, last_sync_on: nil, created_on: nil, updated_on: nil, storage_id: nil, guid: nil, ems_id: nil, last_scan_on: nil, last_scan_attempt_on: nil, uid_ems: nil, retires_on: nil, retired: nil, boot_time: nil, tools_status: nil, standby_action: nil, power_state: nil, state_changed_on: nil, previous_state: nil, connection_state: nil, last_perf_capture_on: nil, blackbox_exists: nil, blackbox_validated: nil, registered: nil, busy: nil, smart: nil, memory_reserve: nil, memory_reserve_expand: nil, memory_limit: nil, memory_shares: nil, memory_shares_level: nil, cpu_reserve: nil, cpu_reserve_expand: nil, cpu_limit: nil, cpu_shares: nil, cpu_shares_level: nil, cpu_affinity: nil, ems_created_on: nil, template: false, evm_owner_id: nil, ems_ref_obj: nil, miq_group_id: nil, linked_clone: nil, fault_tolerance: nil, type: "VmVmware", ems_ref: nil, ems_cluster_id: nil, retirement_warn: nil, retirement_last_warn: nil, vnc_port: nil, flavor_id: nil, availability_zone_id: nil, cloud: false, retirement_state: nil, cloud_network_id: nil, cloud_subnet_id: nil, cloud_tenant_id: nil, raw_power_state: nil, publicly_available: nil, orchestration_stack_id: nil>
[5] pry(main)> v.guest_applications(true)
=> []
[6] pry(main)> v.guest_applications.push([GuestApplication.new, GuestApplication.new])
=> [#<GuestApplication id: nil, name: nil, vendor: nil, version: nil, description: nil, package_name: nil, product_icon: nil, transform: nil, language: nil, typename: nil, vm_or_template_id: nil, product_key: nil, path: nil, arch: nil, host_id: nil, release: nil>,
 #<GuestApplication id: nil, name: nil, vendor: nil, version: nil, description: nil, package_name: nil, product_icon: nil, transform: nil, language: nil, typename: nil, vm_or_template_id: nil, product_key: nil, path: nil, arch: nil, host_id: nil, release: nil>]
[7] pry(main)> v.save!
   (0.2ms)  BEGIN
  SQL (9.4ms)  INSERT INTO "vms" ("autostart", "availability_zone_id", "blackbox_exists", "blackbox_validated", "boot_time", "busy", "cloud", "cloud_network_id", "cloud_subnet_id", "cloud_tenant_id", "config_xml", "connection_state", "cpu_affinity", "cpu_limit", "cpu_reserve", "cpu_reserve_expand", "cpu_shares", "cpu_shares_level", "created_on", "description", "ems_cluster_id", "ems_created_on", "ems_id", "ems_ref", "ems_ref_obj", "evm_owner_id", "fault_tolerance", "flavor_id", "format", "guid", "host_id", "last_perf_capture_on", "last_scan_attempt_on", "last_scan_on", "last_sync_on", "linked_clone", "location", "memory_limit", "memory_reserve", "memory_reserve_expand", "memory_shares", "memory_shares_level", "miq_group_id", "name", "orchestration_stack_id", "power_state", "previous_state", "publicly_available", "raw_power_state", "registered", "retired", "retirement_last_warn", "retirement_state", "retirement_warn", "retires_on", "smart", "standby_action", "state_changed_on", "storage_id", "template", "tools_status", "type", "uid_ems", "updated_on", "vendor", "version", "vnc_port") VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15, $16, $17, $18, $19, $20, $21, $22, $23, $24, $25, $26, $27, $28, $29, $30, $31, $32, $33, $34, $35, $36, $37, $38, $39, $40, $41, $42, $43, $44, $45, $46, $47, $48, $49, $50, $51, $52, $53, $54, $55, $56, $57, $58, $59, $60, $61, $62, $63, $64, $65, $66, $67) RETURNING "id"  [["autostart", nil], ["availability_zone_id", nil], ["blackbox_exists", nil], ["blackbox_validated", nil], ["boot_time", nil], ["busy", nil], ["cloud", false], ["cloud_network_id", nil], ["cloud_subnet_id", nil], ["cloud_tenant_id", nil], ["config_xml", nil], ["connection_state", nil], ["cpu_affinity", nil], ["cpu_limit", nil], ["cpu_reserve", nil], ["cpu_reserve_expand", nil], ["cpu_shares", nil], ["cpu_shares_level", nil], ["created_on", Tue, 20 Jan 2015 22:05:14 UTC +00:00], ["description", nil], ["ems_cluster_id", nil], ["ems_created_on", nil], ["ems_id", nil], ["ems_ref", nil], ["ems_ref_obj", nil], ["evm_owner_id", nil], ["fault_tolerance", nil], ["flavor_id", nil], ["format", nil], ["guid", "6949fd96-a0f0-11e4-b8b6-3c15c2c7fbb8"], ["host_id", nil], ["last_perf_capture_on", nil], ["last_scan_attempt_on", nil], ["last_scan_on", nil], ["last_sync_on", nil], ["linked_clone", nil], ["location", "x"], ["memory_limit", nil], ["memory_reserve", nil], ["memory_reserve_expand", nil], ["memory_shares", nil], ["memory_shares_level", nil], ["miq_group_id", nil], ["name", "x"], ["orchestration_stack_id", nil], ["power_state", nil], ["previous_state", nil], ["publicly_available", nil], ["raw_power_state", nil], ["registered", nil], ["retired", nil], ["retirement_last_warn", nil], ["retirement_state", nil], ["retirement_warn", nil], ["retires_on", nil], ["smart", nil], ["standby_action", nil], ["state_changed_on", nil], ["storage_id", nil], ["template", false], ["tools_status", nil], ["type", "VmVmware"], ["uid_ems", nil], ["updated_on", Tue, 20 Jan 2015 22:05:14 UTC +00:00], ["vendor", "vmware"], ["version", nil], ["vnc_port", nil]]
  SQL (0.3ms)  INSERT INTO "guest_applications" ("arch", "description", "host_id", "language", "name", "package_name", "path", "product_icon", "product_key", "release", "transform", "typename", "vendor", "version", "vm_or_template_id") VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15) RETURNING "id"  [["arch", nil], ["description", nil], ["host_id", nil], ["language", nil], ["name", nil], ["package_name", nil], ["path", nil], ["product_icon", nil], ["product_key", nil], ["release", nil], ["transform", nil], ["typename", nil], ["vendor", nil], ["version", nil], ["vm_or_template_id", 302]]
  SQL (0.2ms)  INSERT INTO "guest_applications" ("arch", "description", "host_id", "language", "name", "package_name", "path", "product_icon", "product_key", "release", "transform", "typename", "vendor", "version", "vm_or_template_id") VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15) RETURNING "id"  [["arch", nil], ["description", nil], ["host_id", nil], ["language", nil], ["name", nil], ["package_name", nil], ["path", nil], ["product_icon", nil], ["product_key", nil], ["release", nil], ["transform", nil], ["typename", nil], ["vendor", nil], ["version", nil], ["vm_or_template_id", 302]]
   (0.3ms)  COMMIT
=> true
```